tracing-subscriber/src/registry/sharded.rs

```diff
@@ -0,0 +1,230 @@
+use sharded_slab::{Guard, Slab};
+
+use crate::{
+    fmt::span::SpanStack,
+    registry::{LookupSpan, SpanData},
+    sync::RwLock,
+};
+use serde_json::{json, Value};
+use std::{
+    cell::RefCell,
+    collections::{HashMap, HashSet},
+    convert::TryInto,
+    fmt, io,
+    marker::PhantomData,
+    sync::{Arc, Mutex},
+};
+use tracing_core::{
+    field::{FieldSet, Visit},
+    span::{self, Id},
+    Event, Field, Interest, Metadata, Subscriber,
+};
+
+pub struct Registry {
+    spans: Arc<Slab<BigSpan>>,
+    // TODO(david): replace this with the span stack from `fmt` (this is wrong)
+    local_spans: RwLock<SpanStack>,
+}
+
+#[derive(Debug)]
+pub struct BigSpan {
+    metadata: &'static Metadata<'static>,
+    parent: Option<Id>,
+    // TODO(david): get rid of these
+    values: Mutex<HashMap<&'static str, Value>>,
+}
+
+// === impl Registry ===
+
+impl Default for Registry {
+    fn default() -> Self {
+        Self {
+            spans: Arc::new(Slab::new()),
+            local_spans: RwLock::new(SpanStack::new()),
+        }
+    }
+}
+
+struct RegistryVisitor<'a>(&'a mut HashMap<&'static str, Value>);
+
+impl<'a> Visit for RegistryVisitor<'a> {
+    // TODO: special visitors for various formats that honeycomb.io supports
+    fn record_debug(&mut self, field: &Field, value: &dyn fmt::Debug) {
+        let s = format!("{:?}", value);
+        self.0.insert(field.name(), json!(s));
+    }
+}
+
+#[inline]
+fn idx_to_id(idx: usize) -> Id {
+    Id::from_u64(idx as u64 + 1)
+}
+
+#[inline]
+fn id_to_idx(id: &Id) -> usize {
+    id.into_u64() as usize - 1
+}
+
+impl Registry {
+    fn insert(&self, s: BigSpan) -> Option<usize> {
+        self.spans.insert(s)
+    }
+
+    fn get(&self, id: &Id) -> Option<Guard<'_, BigSpan>> {
+        self.spans.get(id_to_idx(id))
+    }
+
+    fn take(&self, id: Id) -> Option<BigSpan> {
+        self.spans.take(id_to_idx(&id))
+    }
+}
+
+thread_local! {
+    static CONTEXT: RefCell<SpanStack> = RefCell::new(SpanStack::new());
+}
+
+impl Subscriber for Registry {
+    fn register_callsite(&self, _: &'static Metadata<'static>) -> Interest {
+        Interest::always()
+    }
+
+    fn enabled(&self, _: &Metadata<'_>) -> bool {
+        true
+    }
+
+    #[inline]
+    fn new_span(&self, attrs: &span::Attributes<'_>) -> span::Id {
+        let mut values = HashMap::new();
+        let mut visitor = RegistryVisitor(&mut values);
+        attrs.record(&mut visitor);
+        let parent = attrs.parent().map(|id| id.clone());
```

Comment:
TIOLI:
```suggestion
        let parent = attrs.parent().map(span::Id::clone);
```

Reply:
Also, we need to handle the case where the parent is contextual here. This can be copied from fmt.

Reply:
this needs to call `clone_span` on the parent. Something like: 
```rust
let parent = attrs.parent().map(|id| self.clone_span(id));
```

Reply:
done.